walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,142 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long requestDuration = request.getDuration();
+    Collection<String> requestAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    ArrayList<TimeRange> alreadyBookedTimes = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> alreadyBookedTimesWOptional = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> noOverlap = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> noOverlapWOptional = new ArrayList<TimeRange>();    
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> availableTimesWOptional = new ArrayList<TimeRange>();  
+    int newStart;
+    int newEnd;
+
+    // edge case: if there are no attendees the whole day is returned        
+    if (requestAttendees.isEmpty() && optionalAttendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    // edge case: if the requested meeting is longer than a day there are no options
+    if (requestDuration >= TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();  
+    }
+    alreadyBookedTimes = getAlreadyBookedTimes(events, request);
+    alreadyBookedTimesWOptional = getAlreadyBookedTimesWithOptional(events, request, alreadyBookedTimes);
+    // if there are no meetings the whole day is returned 
+    if (alreadyBookedTimes.isEmpty() && alreadyBookedTimesWOptional.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);    
+    }  
+    noOverlap = removeOverlap(alreadyBookedTimes);
+    noOverlapWOptional = removeOverlap(alreadyBookedTimesWOptional);
+    availableTimesWOptional = addAvailableRanges(request, noOverlapWOptional);
+    availableTimes = addAvailableRanges(request, noOverlap);
+    // return the list of available time ranges
+    if (!availableTimesWOptional.isEmpty()) {
+      return availableTimesWOptional;
+    }
+    return availableTimes;
+  }
+  
+  // if there are multiple meetings consolidate overlapping meetings into one range of time
+  public ArrayList<TimeRange> removeOverlap(ArrayList<TimeRange> times) {
+    ArrayList<TimeRange> noOverlap = new ArrayList<TimeRange>();
+    Collections.sort(times, TimeRange.ORDER_BY_START);
+    int newStart;
+    int newEnd;
+    if (times.size() > 1) {
+      for (int i = 1; i < times.size(); i++) {
+        if (times.get(i - 1).overlaps(times.get(i))) {
+          newStart = Math.min(times.get(i - 1).start(), times.get(i).start());
+          newEnd = Math.max(times.get(i).end(), times.get(i - 1).end());
+          if (!noOverlap.contains(TimeRange.fromStartEnd(newStart, newEnd, false))) {
+            noOverlap.add(TimeRange.fromStartEnd(newStart, newEnd, false));
+          }
+        } else {
+          noOverlap.add(times.get(i - 1));
+          noOverlap.add(times.get(i));
+        } 
+      }      
+    } else {
+      return times;
+    }
+      return noOverlap;
```

Comment:
nit: Let's remove the extra indentation before the `return`.

Reply:
Addressed in "address nits(clarify greeter and its testing, revise variable declaration, fix indentation)"